### ISSUE TITLE
A value of fundingSource of Card payment is string "null" and it causes an exception in BE SDK

### ISSUE THREAD

**Issue by jaejoong-yoo-gtg** [Adyen/adyen-android#1795]
## Description
A value of `fundingSource` of Card payment in a payment data from Android SDK that passes to `onSubmit` of DropInService is string `"null"` instead of `null`. And it causes an exception when we try to deserialise it in BE. The reason is a deserialiser in Adyen BE SDK defines a funding source as enum type, and text "null" couldn't be parsed to any enum.

Please fix Android SDK not to include "null" string as a value. It should have null value or no fundingSource at all.

## Steps to Reproduce
1. I am able to consistently reproduce this issue: Yes
2. Steps to reproduce the issue:
    1. Go to DropIn
    2. Select Card from payment method list
    3. Fill out card information and click on Pay button

## Logs and Crash Reports

Serialised data from `PaymentComponentData.SERIALIZER.serialize(state.data)`
```
{
    "paymentMethod": {
        "type": "scheme",
           .......
        "brand": "visa",
        "threeDS2SdkVersion": "2.2.19",
        "fundingSource": "null"
    },
    .......
}
```

## Code Snippets
Provide code snippets where the SDK is implemented in your project.

```
class PaymentsCheckoutService : DropInService() {

      override fun onSubmit(state: PaymentComponentState<*>) {
          val paymentData = PaymentComponentData.SERIALIZER.serialize(state.data)
          .......
      }
```

## Integration Information
1. Server-side integration: Advanced flow
2. Client-side integration: Drop-in
3. SDK version: com.adyen.checkout:drop-in:5.6.0
4. Android version(s) where issue occurs: Android 14
5. Device model(s) where issue occurs: Pixel 8a

## Additional Context
N/A

**Comment by OscarSpruit** [comment]
Hi @jaejoong-yoo-gtg, thank you for opening an issue! When trying to reproduce it with `PaymentComponentData.SERIALIZER.serialize(state.data)` I noticed that the `JSONObject` does not contain `fundingSource` if it is null, so I think the problem lies somewhere else. Does the response from [`/paymentMethods`](https://docs.adyen.com/api-explorer/Checkout/latest/post/paymentMethods) already contain the `fundingSource` field?

**Comment by jaejoong-yoo-gtg** [comment]
Hi @OscarSpruit, thanks for your reply.

Yes, paymentMethods from our backend contains `fundingSource`.
```
{
  "paymentMethods": [
    {
      "apps": null,
      "brand": null,
      "brands": [
        "amex",
        "mc",
        "visa"
      ],
      "configuration": null,
      "fundingSource": null,
      "group": null,
      "inputDetails": null,
      "issuers": null,
      "name": "Cards",
      "type": "scheme"
    },
......
```

I debugged further related to it and realised that it's caused by PaymentMethods deserialiser.
`PaymentMethodsApiResponse.SERIALIZER.deserialize(it)`

In a `deserialize()` function of `PaymentMethods`, it gets the fundingSource by calling `jsonObject.getStringOrNull(FUNDING_SOURCE)`


This is a snippet of the getStringOrNull extension function.
```
fun JSONObject.getStringOrNull(key: String): String? {
    return if (has(key)) getString(key) else null
}
```

The problem with this function is, it checks if the json object has a key or not, and calls `getString(key)` if it has.
And, `getString()` returns `"null"` if the value is null as it coerces value into a string.

I'm sure you should've called `isNull` instead.
```
return if (isNull(key)) null else getString(key)
```
```
/**
 * Returns true if this object has no mapping for {@code name} or if it has
 * a mapping whose value is {@link #NULL}.
 */
public boolean isNull(@Nullable String name) { ....
```

**Comment by OscarSpruit** [comment]
@jaejoong-yoo-gtg thank you for the detailed investigation! I was able to reproduce it now and we will apply your suggested fix. 

However, even if we apply the fix I think it will not fix your exact problem. Currently, `jsonObject.getStringOrNull(...)` will throw an exception when the value is `null` (exception message: `JSONObject["brand"] is not a string (class org.json.JSONObject$Null : null).`). Because you are not getting this exception I think that on your side you are stringifying `null` when you are converting your backend response to a `JSONObject`. Could you check that?

**Comment by jaejoong-yoo-gtg** [comment]
@OscarSpruit Yes, I already confirmed yesterday if the value in JSONObject is really `null` before passing it to the deserialiser.

I confirmed it with this simple code and also with debugger.
```
suspend fun getPaymentMethods() = repository.getPaymentMethods()?.let {
    Log.e("J", it.toString(4))

    (it.optJSONArray("paymentMethods")?.get(0) as? JSONObject)?.let {
        val fundingSource = it.getStringOrNull("fundingSource")
        Log.e("J", "fundingSource=$fundingSource, is null=${fundingSource == null}")
    }

    PaymentMethodsApiResponse.SERIALIZER.deserialize(it)
}

private fun JSONObject.getStringOrNull(key: String): String? {
    Log.e("J", "is $key null=${isNull(key)}")
    return if (isNull(key)) null else getString(key)
}
```
Above code prints this log
```
... "fundingSource": null, ...

is fundingSource null=true
fundingSource=null, is null=true
```
But if I change a code of `getStringOrNull()` to the SDK's code which is `return if (has(key)) getString(key) else null`, it prints this
```
... "fundingSource": null, ...

is fundingSource null=true
fundingSource=null, is null=false
```

**Comment by jaejoong-yoo-gtg** [comment]
The exception you're experiencing is a bit weird. `getString()` should coerce null to "null" instead of trowing an exception. And, that's what happen to me.

This is how `getString()` designed.
```
/**
 * Returns the value mapped by {@code name} if it exists, coercing it if
 * necessary, or throws if no such mapping exists.
 *
 * @throws JSONException if no such mapping exists.
 */
@NonNull public String getString(@NonNull String name) throws JSONException {
```
I even tested it again with the above test code by changing `getStringOrNull()`.
```
private fun JSONObject.getStringOrNull(key: String): String? {
    Log.e("J", "is $key null=${isNull(key)}")
    return getString(key)
}
```
It returns me `"null"` and prints `fundingSource=null, is null=false` without any exception.

I'm looking at JSONObject packaged in Android API 34. No idea if it has a different behaviour on other versions.

**Comment by OscarSpruit** [comment]
I'm getting this error because of the way I created the `/paymentMethods` response `JSONObject. What I did is:
```Kotlin
JSONObject(
    """
    {
      "apps": null,
      "brand": null,
      "brands": [
        "amex",
        "mc",
        "visa"
      ],
      "configuration": null,
      "fundingSource": null,
      "group": null,
      "inputDetails": null,
      "issuers": null,
      "name": "Cards",
      "type": "scheme"
    }
    """.trimIndent()
)
```
This will put `JSONObject.NULL` in the fields that are `null`, so when `jsonObject.getStringOrNull(...)` tries to cast it to a string it will throw the exception.

This is why I'm curious how you create the `JSONObject`. How does your `repository.getPaymentMethods()` create it?

**Comment by jaejoong-yoo-gtg** [comment]
Hmm. I don't get such an exception at all even with your sample json.
I'm just wondering if you're using `JSONObject` of `org.json` or other packages. I don't see any issues with org.json.JSONObject at least.

What if you try this one?
```
JSONObject("{\"fundingSource\": null}")
```
If there's such an exception, there should have been a lot of crashes reported. I wonder if anyone reported related crashes.

This is a snippet of getPaymentMethods and Retrofit service interface
```
val response = service.getPaymentMethods()
JSONObject(response.string())
```
```
@GET("....")
suspend fun getPaymentMethods(): ResponseBody
```

And, if you provide `@SerializedName` annotation of Gson to PaymentMethodsApiResponse data class, I can just set the data class as a return type and let Retrofit to parse the response without creating JSONObject manually and without using the deserialiser. And, as far as I remember it, Gson ignores fields with null value.
As data class doesn't have such an annotation, it's risky to use it directly as field names can be obfuscated.

**Comment by OscarSpruit** [comment]
The difference we are seeing is because we were using different versions of the org.json library. Previously, I tested this in unit tests were I directly import the latest version of the library. Then when I tried it in our example app I had the exact same behaviour as you. An different version of org.json is apparently shipped with Android.

Anyway, I was now able to verify that your suggested fix does solve the problem. We will be working on it and keep you posted on the progress.
